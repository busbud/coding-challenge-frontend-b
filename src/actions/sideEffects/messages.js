export default {
	requestFailed: 'Request failed',
};
