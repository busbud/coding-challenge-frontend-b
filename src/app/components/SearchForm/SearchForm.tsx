import { addYears, format } from 'date-fns';
import { inject, observer } from 'mobx-react';
import * as React from 'react';
import { I18n } from 'react-i18next';
import styled from 'styled-components';
import { SearchStore } from '../../store/search';

interface Props {
  store: SearchStore;
}

const Form = styled.form`
  margin-top: 50px;
  display: flex;
  flex-wrap: wrap;

  > button {
    margin: 0 5px;
    padding: 10px;
    background: none;
    color: white;
    font-size: 16px;
    border: 1px solid;
  }

  @media (max-width: 920px) {
    justify-content: center;


    > button {
      width: 91.5%;
      margin-top: 15px;
    }
  }
`;

const Input = styled.input`
  margin: 0 5px;
  padding: 10px;
  background: none;
  color: white;
  font-size: 16px;
  border: 1px solid;


  @media (max-width: 920px) {
    width: 45%;
    margin: 5px;
  }
`;

@inject('store')
@observer
class SearchForm extends React.Component<Props> {

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.props.store.clearResults();
    this.props.store.search();
  }

  render() {
    const { setOutboundDate, setPassangerNumber, searchParams } = this.props.store;

    return (
      <I18n ns="">
        {(t) => (
          <Form onSubmit={this.handleSubmit}>
            <Input type="text" id="origin" value={'New York'} onChange={() => {}} />
            <Input type="text" value={t('montreal')} onChange={() => {}} />
            <Input
              type="date"
              min={format(new Date(), 'YYYY-MM-DD')}
              max={format(addYears(new Date(), 1), 'YYYY-MM-DD')}
              value={searchParams.outboundDate}
              onChange={(e) => setOutboundDate(e.target.value)}
            />
            <Input
              type="number"
              value={searchParams.passangerNumber}
              onChange={(e) => setPassangerNumber(Number(e.target.value))}
            />
            <button role="submit">{t('search')}</button>
          </Form>
        )}
     </I18n>
    );
  }
}

export default SearchForm as React.ComponentClass<{}>;
