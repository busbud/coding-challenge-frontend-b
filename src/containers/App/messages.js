import { defineMessages } from 'react-intl';

export default defineMessages({
  error_content: {
    id: 'containers.app.error.content',
    defaultMessage: 'Our apologies an unknown error has occurred.',
  },
});
