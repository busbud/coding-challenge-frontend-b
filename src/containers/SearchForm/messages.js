import { defineMessages } from 'react-intl';

export default defineMessages({
  origin_label: {
    id: 'containers.searchform.origin.label',
    defaultMessage: 'Departure city',
  },
  destination_label: {
    id: 'containers.searchform.destination.label',
    defaultMessage: 'Arrival city',
  },
  outbound_date_label: {
    id: 'containers.searchform.outbound.date.label',
    defaultMessage: 'Date of departure',
  },
});
