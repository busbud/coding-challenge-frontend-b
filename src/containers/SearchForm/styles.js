import red from 'material-ui/colors/red';

export default {
  textField: {
    width: '100%',
  },
  button_wrapper: {
    position: 'relative',
  },
  button: {
    position: 'relative',
    zIndex: 1,
  },
  circularProgress: {
    position: 'absolute',
    top: -8,
    left: -8,
    color: red[500],
  },
};
