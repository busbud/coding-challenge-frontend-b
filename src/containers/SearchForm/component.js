import React from 'react';
import { func, object, bool, instanceOf } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Button from 'material-ui/Button';
import TextField from 'material-ui/TextField';
import { InputAdornment } from 'material-ui/Input';
import SearchIcon from 'material-ui-icons/Search';
import FlightTakeoffIcon from 'material-ui-icons/FlightTakeoff';
import FlightLandIcon from 'material-ui-icons/FlightLand';
import DateRangeIcon from 'material-ui-icons/DateRange';
import { CircularProgress } from 'material-ui/Progress';
import { DatePicker } from 'material-ui-pickers';
import messages from './messages';
import Grid from 'material-ui/Grid';

const propTypes = {
  classes: object.isRequired,
  onSubmit: func.isRequired,
  onChange: func.isRequired,
  outbound_date: instanceOf(Date).isRequired,
  isFetching: bool.isRequired,
};

const Component = ({ classes, outbound_date, onChange, onSubmit, isFetching }) => (
  <form onSubmit={onSubmit}>
    <Grid container justify="center">
      <Grid item md xs={12}>
        <TextField
          id="origin"
          className={classes.textField}
          label={<FormattedMessage {...messages.origin_label} />}
          defaultValue="New York"
          InputProps={{
            readOnly: true,
            startAdornment: (
              <InputAdornment position="start">
                <FlightTakeoffIcon />
              </InputAdornment>
            ),
          }}
        />
      </Grid>
      <Grid item md xs={12}>
        <TextField
          id="destination"
          className={classes.textField}
          label={<FormattedMessage {...messages.destination_label} />}
          defaultValue="Montreal"
          InputProps={{
            readOnly: true,
            startAdornment: (
              <InputAdornment position="start">
                <FlightLandIcon />
              </InputAdornment>
            ),
          }}
        />
      </Grid>

      <Grid item md xs={12}>
        <DatePicker
          id="outbound_date"
          className={classes.textField}
          label={<FormattedMessage {...messages.outbound_date_label} />}
          value={outbound_date}
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <DateRangeIcon />
              </InputAdornment>
            ),
          }}
          onChange={onChange}
          disablePast
        />
      </Grid>

      <div className={classes.button_wrapper}>
        <Button type="submit" variant="fab" className={classes.button} color="primary" aria-label="search">
          <SearchIcon />
        </Button>
        {isFetching && <CircularProgress className={classes.circularProgress} size={72} />}
      </div>
    </Grid>
  </form>
);

Component.propTypes = propTypes;

export default Component;
