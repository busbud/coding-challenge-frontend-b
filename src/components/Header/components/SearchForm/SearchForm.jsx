import React, { useState } from "react";
import PropTypes from "prop-types";
import { useTranslation } from "react-i18next";

import Select from "commons/Select";
import DatePicker from "commons/DatePicker";
import Button from "commons/Button";

import { Form, FormControl } from "./styles";

const SearchForm = ({ onSearch, ...otherProps }) => {
  const { t } = useTranslation();
  const [originCity, setOriginCity] = useState("");
  const [destinationCity, setDestinationCity] = useState("");
  const [passengers, setPassengers] = React.useState("");
  const [outboundDate, setOutboundDate] = React.useState(null);

  const handleOriginCityChange = (value) => {
    setOriginCity(value);
  };

  const handleDestinationCityChange = (value) => {
    setDestinationCity(value);
  };

  const handlePassengersChange = (value) => {
    setPassengers(value);
  };

  const handleDateChange = (date, dateString) => {
    setOutboundDate(dateString);
  };

  const handleSearch = () => {
    onSearch({
      originCity,
      destinationCity,
      outboundDate,
      passengers,
    });
  };

  return (
    <Form {...otherProps}>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={originCity}
          onChange={handleOriginCityChange}
          data-testid="origin-city-select"
        >
          <Select.Option value="">
            <em>{t("originCity")}</em>
          </Select.Option>
          <Select.Option value="f2m673">Québec</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={destinationCity}
          onChange={handleDestinationCityChange}
          data-testid="destination-city-select"
        >
          <Select.Option value="">
            <em>{t("destinationCity")}</em>
          </Select.Option>
          <Select.Option value="f25dvk">Montréal</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={passengers}
          onChange={handlePassengersChange}
          data-testid="passengers-select"
        >
          <Select.Option value="">
            <em>{t("passengers")}</em>
          </Select.Option>
          <Select.Option value="1">1 {t("adult")}</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <DatePicker
          style={{ width: "100%" }}
          onChange={handleDateChange}
          data-testid="outbound-date-picker"
          placeholder="Outbound Date"
        />
      </FormControl>
      <FormControl>
        <Button
          variant="contained"
          color="primary"
          onClick={handleSearch}
          data-testid="search-button"
        >
          {t("search")}
        </Button>
      </FormControl>
    </Form>
  );
};

SearchForm.propTypes = {
  /**
   * The search event trigger.
   */
  onSearch: PropTypes.func.isRequired,
};

export default SearchForm;
