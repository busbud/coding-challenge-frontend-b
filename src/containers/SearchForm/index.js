import { compose } from 'recompose';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { format } from 'date-fns';
import { withStyles } from 'material-ui/styles';
import { makeSelectFetchingState, makeSelectOutboundDate } from '../App/selectors';
import { updateQuery, fetchDepartures } from '../App/actions';
import styles from './styles';
import component from './component';

export const mapDispatchToProps = (dispatch, props) => {
  return {
    onChange(date) {
      const outbound_date = format(new Date(date), 'YYYY-MM-DD');
      dispatch(updateQuery({ path: { outbound_date } }));
    },
    onSubmit(event) {
      if (event) event.preventDefault();
      dispatch(fetchDepartures());
    },
  };
};
export const mapStateToProps = createStructuredSelector({
  isFetching: makeSelectFetchingState(),
  outbound_date: makeSelectOutboundDate(),
});
export const enhance = compose(withStyles(styles), connect(mapStateToProps, mapDispatchToProps));

export default enhance(component);
