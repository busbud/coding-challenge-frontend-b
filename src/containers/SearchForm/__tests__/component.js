import React from 'react';
import { shallow } from 'enzyme';
import TextField from 'material-ui/TextField';
import { DatePicker } from 'material-ui-pickers';
import SearchForm from '../component';
import messages from '../messages';
import { toDate } from 'date-fns';

const props = {
  classes: {
    root: 'fakeRootClass',
    textField: 'fakeTextfieldClass',
    button_wrapper: 'fakeButtonWrapperClass',
    button: 'fakeButtonClass',
    circularProgress: 'fakeCircularProgressClass',
  },
  onSubmit: jest.fn(),
  onChange: jest.fn(),
  outbound_date: toDate('2018-08-02'),
  isFetching: true,
};

describe('component | SearchForm | component', () => {
  describe('render', () => {
    it('should render without throwing and error', () => {
      // given
      const wrapper = shallow(<SearchForm {...props} />);

      // then
      expect(wrapper).toBeDefined();
    });

    it('should match snapshot', () => {
      // given
      const wrapper = shallow(<SearchForm {...props} />);

      // then
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('form', () => {
    it('should have proper onSubmit props', () => {
      // given
      const wrapper = shallow(<SearchForm {...props} />);
      const { onSubmit } = wrapper.props();

      // then
      expect(onSubmit).toBe(props.onSubmit);
    });
  });

  describe('Field', () => {
    describe('origin', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const { className } = wrapper
          .find(TextField)
          .filter('#origin')
          .props();

        // then
        expect(className).toBe(props.classes.textField);
      });
      it('should have proper label', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const formattedMessageProps = wrapper
          .find(TextField)
          .filter('#origin')
          .props().label.props;

        // then
        expect(formattedMessageProps).toMatchObject(messages.origin_label);
      });
    });
    describe('destination', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const { className } = wrapper
          .find(TextField)
          .filter('#destination')
          .props();

        // then
        expect(className).toBe(props.classes.textField);
      });
      it('should have proper label', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const formattedMessageProps = wrapper
          .find(TextField)
          .filter('#destination')
          .props().label.props;

        // then
        expect(formattedMessageProps).toMatchObject(messages.destination_label);
      });
    });
    describe('outbound_date', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const { className } = wrapper
          .find(DatePicker)
          .filter('#outbound_date')
          .props();

        // then
        expect(className).toBe(props.classes.textField);
      });
      it('should have proper label', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const formattedMessageProps = wrapper.find(DatePicker).props().label.props;

        // then
        expect(formattedMessageProps).toMatchObject(messages.outbound_date_label);
      });
      it('should have proper value', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const { value } = wrapper.find(DatePicker).props();

        // then
        const expectedValue = props.outbound_date;
        expect(value).toEqual(expectedValue);
      });
      it('should have correct onChange props', () => {
        // given
        const wrapper = shallow(<SearchForm {...props} />);
        const { onChange } = wrapper.find(DatePicker).props();

        // then
        expect(onChange).toEqual(props.onChange);
      });
    });
  });
});
