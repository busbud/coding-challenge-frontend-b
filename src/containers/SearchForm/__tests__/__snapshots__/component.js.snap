// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component | SearchForm | component render should match snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <WithStyles(GridWrapper)
    container={true}
    justify="center"
  >
    <WithStyles(GridWrapper)
      item={true}
      md={true}
      xs={12}
    >
      <TextField
        InputProps={
          Object {
            "readOnly": true,
            "startAdornment": <WithStyles(InputAdornment)
              position="start"
          >
              <pure(FlightTakeoff) />
          </WithStyles(InputAdornment)>,
          }
        }
        className="fakeTextfieldClass"
        defaultValue="New York"
        id="origin"
        label={
          <FormattedMessage
            defaultMessage="Departure city"
            id="containers.searchform.origin.label"
            values={Object {}}
          />
        }
        required={false}
        select={false}
      />
    </WithStyles(GridWrapper)>
    <WithStyles(GridWrapper)
      item={true}
      md={true}
      xs={12}
    >
      <TextField
        InputProps={
          Object {
            "readOnly": true,
            "startAdornment": <WithStyles(InputAdornment)
              position="start"
          >
              <pure(FlightLand) />
          </WithStyles(InputAdornment)>,
          }
        }
        className="fakeTextfieldClass"
        defaultValue="Montreal"
        id="destination"
        label={
          <FormattedMessage
            defaultMessage="Arrival city"
            id="containers.searchform.destination.label"
            values={Object {}}
          />
        }
        required={false}
        select={false}
      />
    </WithStyles(GridWrapper)>
    <WithStyles(GridWrapper)
      item={true}
      md={true}
      xs={12}
    >
      <withUtilsDatePickerWrapper
        InputProps={
          Object {
            "startAdornment": <WithStyles(InputAdornment)
              position="start"
          >
              <pure(DateRange) />
          </WithStyles(InputAdornment)>,
          }
        }
        className="fakeTextfieldClass"
        disablePast={true}
        id="outbound_date"
        label={
          <FormattedMessage
            defaultMessage="Date of departure"
            id="containers.searchform.outbound.date.label"
            values={Object {}}
          />
        }
        onChange={[Function]}
        value={2018-08-02T00:00:00.000Z}
      />
    </WithStyles(GridWrapper)>
    <div
      className="fakeButtonWrapperClass"
    >
      <WithStyles(Button)
        aria-label="search"
        className="fakeButtonClass"
        color="primary"
        type="submit"
        variant="fab"
      >
        <pure(Search) />
      </WithStyles(Button)>
      <WithStyles(CircularProgress)
        className="fakeCircularProgressClass"
        size={72}
      />
    </div>
  </WithStyles(GridWrapper)>
</form>
`;
